### PR TITLE
Mavlink threadding fixes

### PR DESCRIPTION
**Pull request by bkueng**
This fixes point 5 from #7223 by moving the mavlink mission, param manager, ftp & log downloader into the receiver thread.

Tested on:
- SITL
- Pixracer, USB & WiFi
- Pixhawk with 3DR telemetry

Test cases & results:
- Param loading: pretty much the same observable behavior as before
- Large mission upload & download (ca. 400WP): transaction times are roughly the same before and after
- Log download: download speed is a bit faster now via USB, unchanged via telemetry (at 1.5KB/s, not that I think someone would actually do that :). Looking at the mavlink inspector, the attitude rate goes down to ~45Hz from 50Hz during download, which is expected. I get the same numbers on current master.

I also tested ULog streaming + log download in parallel via USB, and it works well, the attitude was still received with ~40Hz.

@lovettchris can you test if you still have FTP issues with this?
@ecmnet @tops4u can you test as well?

### CODE REVIEW

**Comment by tops4u** [comment]
I'll be flying Missions on the Weekend currently on an RC Build a few weeks old. I don't have my own Build infrastructure so I'll have to wait for a Master Build being available for Download.

**Comment by dagar** [comment]
@tops4u https://s3.amazonaws.com/px4-travis/Firmware/mavlink_threadding_fixes/px4fmu-v2_default.px4

**Comment by ecmnet** [comment]
@bkueng I just tested your branch. No significant change compared to master in my case. Lost package-rate with companion connection at 921600 baud is about 1,5 packages per second. This increases significantly when reading a log via Mavlink (slightly faster, ~10 Packages per second, 5MB file read at a rate of **14kb/s**). Streaming ULOG returns a ULOG package lost ratio of ~12% (similar with master). But I never experienced a MAVlink freeze  with my setup.

**Comment by bkueng** [comment]
@ecmnet what is your mavlink rate set to (or SYS_COMPANION param)? What does `mavlink status` say with and without ULog streaming?

**Comment by ecmnet** [comment]
Without ULOG streaming:

a

with ULOG streaming:

b


Note, this is MAVlinkShell. Second mavlink status reply seems to be messed up a bit

**Comment by bkueng** [comment]
You can see `txerr: 14.778` which explains the high drop rate. This either means the physical link is saturated (or limited by the mavlink rate, but from the output I conclude it's set to 80k), or more likely the NuttX TX buffer is full. You can try to increase it, it's one of the `CONFIG_USARTx_TXBUFSIZE` configs, for example [here](https://github.com/PX4/Firmware/blob/master/nuttx-configs/px4fmu-v2/nsh/defconfig#L958), I don't know which UART number it is in your case (do a `make clean` every time you change it).
Also the ULog rate is quite high, it could occasionally hit the limit of 70%, which would also explain ULog drops.

**Comment by ecmnet** [comment]
Ok thanks, I'll give that a try.

**Comment by ecmnet** [comment]
@bkueng I had a significant improvement switching the buffersizexs of USART3 to 600 for rx and 1500 for tx. ULOG has (nearly) no drops anymore. Reading a log via mavlink still has some drops, but much less than before. My tests were based on your branch:

c

**Comment by bkueng** [comment]
Thanks for your extensive testing! Is increasing the RX buffer needed as well?
The thing is that increasing the buffers directly reduces the free RAM by the same amount (it's a fixed allocation), so I'd like to keep it as small as possible.
I think we could increase TX to something between 1500 and 2000, but I leave the last call on this up to @LorenzMeier.

@LorenzMeier What do you think about including this PR into the release?

**Comment by ecmnet** [comment]
@bkueng I increased the RX buffer to 600 (as it was set for USART2). I did not perform any tests on RX side.

**Comment by LorenzMeier** [comment]
Probably makes sense to get this into the release for the stability sake.

File: src/modules/mavlink/mavlink_mission.cpp [hunk 1]
```diff
@@ -101,20 +102,6 @@ MavlinkMissionManager::~MavlinkMissionManager()
 	orb_unadvertise(_offboard_mission_pub);
 }
 
-unsigned
-MavlinkMissionManager::get_size()
```
Comment (LorenzMeier) [inline src/modules/mavlink/mavlink_mission.cpp]: Why did you remove this? Doesn't this lead to buffer saturation when mission transfers are active?

File: src/modules/mavlink/mavlink_mission.cpp [hunk 2]
```diff
@@ -86,8 +86,9 @@ MavlinkMissionManager::MavlinkMissionManager(Mavlink *mavlink) : MavlinkStream(m
 	_offboard_mission_sub(-1),
 	_mission_result_sub(-1),
 	_offboard_mission_pub(nullptr),
-	_slow_rate_limiter(_interval / 5.0f),
-	_verbose(false)
+	_slow_rate_limiter(100 * 1000), // Rate limit sending of the current WP sequence to 10 Hz
```
Comment (LorenzMeier) [inline src/modules/mavlink/mavlink_mission.cpp]: This is not desirable as some links will get heartbeat timeouts if you fill them with mission items while other fast links will take many seconds for a couple hundred mission items even though they could transfer much faster.

File: src/modules/mavlink/mavlink_parameters.cpp [hunk 3]
```diff
@@ -319,15 +314,19 @@ MavlinkParametersManager::handle_message(const mavlink_message_t *msg)
...
-		int i = 0;
-
-		while (i++ < 5 && send_one());
+		// speed up parameter loading via UDP, TCP or USB: try to send 15 at once
+		max_num_to_send = 5 * 3;
 	}
+
+	int i = 0;
+
+	while (i++ < max_num_to_send && send_one());
```
Comment (LorenzMeier) [inline src/modules/mavlink/mavlink_parameters.cpp]: That makes sense.

File: src/modules/mavlink/mavlink_mission.cpp (hunk 2 above)
Comment (bkueng) [inline src/modules/mavlink/mavlink_mission.cpp]: This rate limiter is only applied for sending the current waypoint. i.e. during a mission, when the vehicle reaches a waypoing, this will change, and QGC can update the currently marked waypoint.

File: src/modules/mavlink/mavlink_mission.cpp (hunk 1 above)
Comment (bkueng) [inline src/modules/mavlink/mavlink_mission.cpp]: Since the object is now entirely in the receiver thread, the main thread does not know about it anymore, and thus does not include it in the rate calculations.
I was a bit concerned about this too, but my tests showed that it still works well (with fast and slow links).
With regard to https://github.com/PX4/Firmware/issues/7250, we could generally reduce the streams, when a mission/param transfer or log download is in progress.
